File: typescript-examples/browser-sdk-showcase/api/ai/extract-structured-data.ts
// https://docs.intunedhq.com/automation-sdks/intuned-sdk/typescript/ai/functions/extractStructuredData
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";
import { extractStructuredData } from "@intuned/browser/ai";
import { z } from "zod";

interface Params {
  // No params needed
}

const BookSchema = z.object({
  name: z.string().describe("Book title"),
  price: z.string().describe("Book price"),
  description: z.string().nullable().describe("Book description"),
  in_stock: z.boolean().describe("Stock availability"),
  rating: z.string().nullable().describe("Book rating"),
});

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
) {
  await goToUrl({
    page,
    url: "https://books.toscrape.com/catalogue/a-light-in-the-attic_1000/index.html",
  });

  // Extract from the Page directly using Zod schema.
  // You can also extract from a specific locator or by passing TextContentItem.
  const product = await extractStructuredData({
    source: page,
    strategy: "HTML",
    model: "gpt-5-mini",
    dataSchema: BookSchema, // Pass Zod schema directly, or use a normal JSON schema object too.
    prompt: "Extract book details from this page",
    enableCache: false, // To enable cache, you must run in Intuned context (IDE/CLI) and save the project, with the correct API credentials.
    maxRetries: 3,
  });

  console.log(`Found product: ${product.name} - ${product.price}`);

  return {
    name: product.name,
    price: product.price,
    description: product.description,
    in_stock: product.in_stock,
    rating: product.rating,
  };
}

File: python-examples/browser-sdk-showcase/api/ai/extract-structured-data.py
# https://docs.intunedhq.com/automation-sdks/intuned-sdk/python/ai/functions/extract_structured_data
from playwright.async_api import Page
from typing import TypedDict
from intuned_browser import go_to_url
from pydantic import BaseModel, Field
from intuned_browser.ai import extract_structured_data


class Params(TypedDict):
    pass


async def automation(page: Page, params: Params | None = None, **_kwargs):
    class Book(BaseModel):
        name: str = Field(description="Book title")
        price: str = Field(description="Book price")
        description: str | None = Field(default=None, description="Book description")
        in_stock: bool = Field(description="Stock availability")
        rating: str | None = Field(default=None, description="Book rating")

    await go_to_url(
        page,
        "https://books.toscrape.com/catalogue/a-light-in-the-attic_1000/index.html",
    )

    # Extract from the Page directly using Pydantic model.
    # You can also extract from a specific locator or by passing TextContentItem.
    # Check https://docs.intunedhq.com/automation-sdks/intuned-sdk/python/helpers/functions/extract_structured_data for more details.
    product = await extract_structured_data(
        source=page,
        strategy="HTML",
        model="gpt-5-mini",
        data_schema=Book,  # Pass Pydantic model directly, or use a normal json schema dictionary too.
        prompt="Extract book details from this page",
        enable_cache=False,  # To enable cache, you must run in Intuned context (IDE/CLI) and save the project, with the correct API credentials.
        max_retries=3,
    )
    print(f"Found product: {product['name']} - {product['price']}")

    return "Success"
